[Discover] Data selector enhancement

mengweieric opened the pull request:
### Description

Discover data selector enhancements includes:

- Add a few missing i18n
- Add id, improve data source metadata with fields addition for data sources
- Improve data source filtering
- Add in-memory cache for data selector list with refresh button
- Improve data selector rendering logics
- Removed hard coded data source type-to-display mapping from selector and read from metadata
- Increased test coverage
- Types and interfaces improvements

### Issues Resolved

- https://github.com/opensearch-project/OpenSearch-Dashboards/issues/4991
- https://github.com/opensearch-project/OpenSearch-Dashboards/issues/5211

<!-- Example: closes #1234 or fixes <Issue_URL> -->

## Screenshot
<img width="658" alt="Screenshot 2024-04-29 at 12 03 46 PM" src="https://github.com/opensearch-project/OpenSearch-Dashboards/assets/80358241/ae95e641-1db2-4cf5-8e41-a6e7576ea896">

## Testing the changes

<!--
  Please provide detailed steps for validating your changes. This could involve specific commands to run,
  pages to visit, scenarios to try or any other information that would help reviewers verify
  the functionality of your change
-->

## Changelog
<!--
Add a short but concise sentence about the impact of this pull request. Prefix an entry with the type of change they correspond to: breaking, chore, deprecate, doc, feat, fix, infra, refactor, test.
- fix: Update the graph
- feat: Add a new feature

If this change does not need to added to the changelog, just add a single `skip` line e.g.
- skip

Descriptions following the prefixes must be 100 characters long or less
-->

### Check List

- [x] All tests pass
  - [x] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
- [x] New functionality includes testing.
- [x] New functionality has been documented.
- [x] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

kavilla commented on src/plugins/data/public/data_sources/register_default_datasource.ts:
```diff
@@ -4,23 +4,32 @@
+        ui: {
+          label: 'Index patterns', // display name of your data source,
+          typeLabel: 'Index patterns', // display name of your data source type,
+          typeGroup: 'DEFAULT_INDEX_PATTERNS',
```
i think this can be an enum

mengweieric replied:
Thanks for early review, thinking about the same, will update it.

kavilla reviewed (dismissed):
Really like these changes. Just have some nits but I don't see any reason to block this unless CI fails.

kavilla commented on changelogs/fragments/6571.yml:
```diff
@@ -0,0 +1,2 @@
+refactor:
+- discover data selector enhancement and refactoring ([#6571](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/6571))
```
Can we verify it's just Discover? I think it's any thing that uses it the data source selector right but just doesn't pass it.

mengweieric replied:
For now it's just discover, but since we have already aligned on having data selector as its name to distinguish from MDS selector I think we can remove 'discover' as it can be used outside of discover.

kavilla replied:
gotcha.

but also if some 3rd party plugin referenced this selector technically it would be modified as well right?

mengweieric replied:
yea that is correct.

kavilla replied:
sounds good to me! thanks!

kavilla commented on src/plugins/data/public/data_sources/constants.tsx:
```diff
@@ -0,0 +1,29 @@
+import { DataSourceUIGroupType } from './datasource/types';
+import { DEFAULT_DATA_SOURCE_DISPLAY_NAME } from './register_default_datasource';
+
+export const defaultDataSourceMetadata = {
```
qq: did we settle on this code wise being internally referenced as data source?

mengweieric replied:
In last Thursday's meeting @kgcreative mentioned we should for now only aligns on data connector in replace of previous data source selector naming, but not for renaming other existing datasource names through out the code for this. Soon will have a push for renaming stuff only for datasource selector to be data selector.

kavilla commented:
one more question, is this configurable with a ui setting? or config? could be an easy win to let it be a ui setting

mengweieric commented on src/plugins/data/public/data_sources/datasource_selector/datasource_selectable.tsx:
```diff
@@ -162,6 +165,24 @@ export const DataSourceSelectable = ({
       singleSelection={singleSelection}
       isClearable={false}
+      append={
+        <EuiText size="s">
```
Will add this with a push soon.

mengweieric commented:
> one more question, is this configurable with a ui setting? or config? could be an easy win to let it be a ui setting

for now is not configurable through ui setting, but I believe what you are pointing to is the ideal way we want to go with in near releases.

joshuali925 commented on src/plugins/data/public/data_sources/datasource/types.ts:
```diff
@@ -3,49 +3,79 @@
 }
 
-export interface IDataSourceGroup {
+export interface DataSetWithDataSource<T = []> {
+  ds: DataSource;
+  list: T;
+}
```
maybe
```suggestion
export interface DataSetWithDataSource<T = unknown> {
  ds: DataSource;
  list: T[];
}
```

joshuali925 commented on src/plugins/data/public/data_sources/datasource/types.ts:
```diff
@@ -3,49 +3,79 @@
+  query: T;
+}
+
+export interface IDataSourceQueryResult<T = {}> {
```
why are the generics defaulting to `{}`?

mengweieric replied:
For these interfaces basically there's no concrete implementations needed at this point so just providing a default object so elsewhere it doesn't need to specify a type which it doesn't really need.

joshuali925 commented on src/plugins/data/public/data_sources/datasource/types.ts:
```diff
@@ -3,49 +3,79 @@
+}
+
+export interface DataSourceConnectionStatus {
+  status: string;
```
can the status be an arbitrary string or can the type be more narrowed down?

joshuali925 commented on src/plugins/data/public/data_sources/datasource/types.ts:
```diff
@@ -3,49 +3,79 @@
+  typeLabel: string; // the display name of data source type
+  displayOrder?: number; // the order in which the data source should be displayed in selector
+  description?: string; // short description of your database
+  icon?: string; // uri of the icon
```
i think tsdoc/jsdoc has `@member` or `/** xxx */` to document fields of interface, might be better so typescript language server would be aware of it

i'm not sure if `// xxx` works, if it does then ignore

kavilla reviewed (approved):
Link is not failing because of this PR. All good to me

ashwin-pc commented on CHANGELOG.md:
```diff
@@ -108,6 +108,7 @@ Inspired from [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 - [Workspace] Add workspaces filter to saved objects page. ([#6458](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/6458))
 - [Multiple Datasource] Support multi data source in Region map ([#6654](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/6654))
 - Add `rightNavigationButton` component in chrome service for applications to register and add dev tool to top right navigation. ([#6553](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/6553))
+- [Discover Data Selector] - Data selector enhancement and refactoring ([#6571](https://github.com/opensearch-project/OpenSearch-Dashboards/pull/6571))
```
We dont need this in two places. You can delete this.

ashwin-pc commented on src/plugins/data/public/data_sources/constants.tsx:
```diff
@@ -0,0 +1,46 @@
+import { i18n } from '@osd/i18n';
+import { DataSourceUIGroupType } from './datasource/types';
+
+export const S3_GLUE_DATA_SOURCE_DISPLAY_NAME = 'Amazon S3';
```
Are we still hard coding these? Is there a plan to remove them?

ashwin-pc commented on src/plugins/data/public/data_sources/datasource_services/datasource_service.ts:
```diff
@@ -50,21 +56,24 @@ export class DataSourceService {
-    const dsName = ds.getName();
-    if (dsName in this.dataSources) {
+  async registerDataSource(ds: DataSource): Promise<IDataSourceRegistrationResult> {
+    const dsId = ds.getId();
```
Nit: dont use abbreviated variable names. Always expand them.

ashwin-pc commented on src/plugins/data/public/data_sources/datasource/types.ts:
```diff
@@ -3,49 +3,112 @@
+  data: T;
+}
+
+export enum DataSourceUIGroupType {
```
Same as above, do we have a plan on getting away from hardcoding datasource types in this service? These should be generic enough to allow different groups of datasources.

ashwin-pc reviewed (commented):
Can you outline in the PR description what exactly this change is about? Currently it looks like this addresses all the feedback from the previous PR and closes the Data Sources issue. However i see in here that S3 and Spark data sources are still hard coded and the registration function is still not generic enough to handle future data sources.

mengweieric commented:
> Can you outline in the PR description what exactly this change is about? Currently it looks like this addresses all the feedback from the previous PR and closes the Data Sources issue. However i see in here that S3 and Spark data sources are still hard coded and the registration function is still not generic enough to handle future data sources.

Thanks for the review.

The removal of hardcoding in this PR is to remove previous hardcoded type-to-display-name mapping and instead reading from data source metadata. The s3 data sources are backend data sources where it doesn't have any UI specific definitions. Ideally these front end specific definitions should be retrieving from settings or front end data sources separated from backend data sources where user can configure them. Does this align with the direction you are pointing to? 

As for registration function, are you point to the registerDataSources in data source service or the way data source owner register the data source?

kavilla commented:
> > Can you outline in the PR description what exactly this change is about? Currently it looks like this addresses all the feedback from the previous PR and closes the Data Sources issue. However i see in here that S3 and Spark data sources are still hard coded and the registration function is still not generic enough to handle future data sources.
> 
> Thanks for the review.
> 
> The removal of hardcoding in this PR is to remove previous hardcoded type-to-display-name mapping and instead reading from data source metadata. The s3 data sources are backend data sources where it doesn't have any UI specific definitions. Ideally these front end specific definitions should be retrieving from settings or front end data sources separated from backend data sources where user can configure them. Does this align with the direction you are pointing to?
> 
> As for registration function, are you point to the registerDataSources in data source service or the way data source owner register the data source?

I think @ashwin-pc is saying the next iteration of this where if I am a cluster admin, I can install a plugin that adds the new data source. For example, if I want a S3 connection type I install a S3 connection plugin that wires that up. Or better yet, some type of catalogue.

mengweieric commented:
> > > Can you outline in the PR description what exactly this change is about? Currently it looks like this addresses all the feedback from the previous PR and closes the Data Sources issue. However i see in here that S3 and Spark data sources are still hard coded and the registration function is still not generic enough to handle future data sources.
> > 
> > 
> > Thanks for the review.
> > The removal of hardcoding in this PR is to remove previous hardcoded type-to-display-name mapping and instead reading from data source metadata. The s3 data sources are backend data sources where it doesn't have any UI specific definitions. Ideally these front end specific definitions should be retrieving from settings or front end data sources separated from backend data sources where user can configure them. Does this align with the direction you are pointing to?
> > As for registration function, are you point to the registerDataSources in data source service or the way data source owner register the data source?
> 
> I think @ashwin-pc is saying the next iteration of this where if I am a cluster admin, I can install a plugin that adds the new data source. For example, if I want a S3 connection type I install a S3 connection plugin that wires that up. Or better yet, some type of catalogue.

Yea understood that we will want this in future iterations, this PR addressed the step one mapping (updated the description), eventually we will not want those to be defined in code.

ashwin-pc commented:
@mengweieric yes my goal is to have no service specific names hardcoded in the framework. If any other company wanted to extend the datasources to support their version of the service like Azure for example. They should not need to make a code change in OSD to show up as a new Datasource as long as they adhere to the schema.

e.g. 

```
registerDataSource({
  name: 'Postgress`
  schema: any
  ...
})
```

today, if i wanted to add postgress or any other datasource type to this list, i will need to make a code change in OSD to show up as an option in the DataSources menu. That should not be necessary. That should all be abstracted away in the API.
